### DISCUSSION TITLE
Best practise for writing business logic

### DISCUSSION THREAD

**Discussion by umair975** [directus/directus#12778] [Archive]
Looking at moving laravel/mysql based web application to a headless CMS environment. This should give us the advantage of having data organized and served via REST/GraphQL endpoints. Most probably we will connect next js/ react js frontend and reactnative/flutter for mobile apps. What I am looking for is suggestion on how you manage business logic? at what level? things like writing data to multiple tables (master-detail), calculating rates, validating applicable promotions etc. Is it possible to do these on Directus? custom endpoints? or these needs to be handled elsewhere.

**Comment by rijkvanzanten** [comment]
It's mostly personal preference! If you have small pieces of logic, you might prefer using hooks/endpoints to keep things within Directus. If you have a lot of custom logic that needs other specific integrations or has specific tech stack requirements, you might prefer using a standalone micro service, and use Directus as the data source
